// nestjs-zod/dist/dto.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_symbols = require('./symbols-uv4cJ8eW.cjs');
let zod_v4_core = require("zod/v4/core");

//#region src/dto.ts
function createZodDto(schema, options) {
	class AugmentedZodDto {
		static {
			this.isZodDto = true;
		}
		static {
			this.schema = schema;
		}
		static {
			this.codec = options?.codec || false;
		}
		static {
			this[require_symbols.ioSymbol] = "input";
		}
		static create(input) {
			return this.schema.parse(input);
		}
		static get Output() {
			require_symbols.assert("_zod" in schema, "Output DTOs can only be created from zod v4 schemas");
			class AugmentedZodDto {
				static {
					this.isZodDto = true;
				}
				static {
					this.schema = schema;
				}
				static {
					this[require_symbols.ioSymbol] = "output";
				}
				static create(input) {
					return this.schema.parse(input);
				}
				static _OPENAPI_METADATA_FACTORY() {
					return openApiMetadataFactory({
						schema: this.schema,
						io: "output"
					});
				}
			}
			Object.defineProperty(AugmentedZodDto, "name", { value: `${this.name}_Output` });
			return AugmentedZodDto;
		}
		static _OPENAPI_METADATA_FACTORY() {
			return openApiMetadataFactory({
				schema: this.schema,
				io: "input"
			});
		}
	}
	return AugmentedZodDto;
}
function openApiMetadataFactory({ schema, io }) {
	if (!("_zod" in schema) && "_def" in schema && io === "output") throw new Error("[nestjs-zod] Output schemas are not supported for zod@v3");
	if (!("_zod" in schema) && !("_def" in schema)) return {};
	const { $defs, $schema: _$schema, ...generatedJsonSchema } = generateJsonSchema(schema, io);
	/**
	* nestjs expects us to return a record of properties
	*
	* However, in some cases, we can't return a record of properties.  For
	* example, arrays, intersections, and unions can not be represented like this
	*
	* As a workaround, we wrap the schema in a "root" object.  Then in the
	* `cleanupOpenApiDoc` function, we unwrap the root object.
	*/
	const jsonSchema = !isObjectTypeWithProperties(generatedJsonSchema) ? {
		type: "object",
		id: generatedJsonSchema.id,
		title: generatedJsonSchema.title,
		properties: { root: {
			...generatedJsonSchema,
			[require_symbols.UNWRAP_ROOT_KEY]: true
		} },
		$defs
	} : {
		...generatedJsonSchema,
		$defs
	};
	const { hasRefs, hasNull, hasConst } = getSchemaMetadata(jsonSchema);
	const properties = {};
	for (const [propertyKey, propertySchema] of Object.entries(jsonSchema.properties || {})) {
		const newPropertySchema = {
			...propertySchema,
			type: propertySchema.type || ""
		};
		if (hasConst) newPropertySchema[require_symbols.HAS_CONST_KEY] = true;
		if (hasNull) newPropertySchema[require_symbols.HAS_NULL_KEY] = true;
		if (hasRefs) newPropertySchema[require_symbols.PARENT_HAS_REFS_KEY] = true;
		if (typeof propertySchema.type !== "string") newPropertySchema[require_symbols.EMPTY_TYPE_KEY] = true;
		const required = Boolean("required" in jsonSchema && jsonSchema.required?.includes(propertyKey));
		if (newPropertySchema.type === "object") {
			newPropertySchema.selfRequired = required;
			newPropertySchema[require_symbols.SELF_REQUIRED_KEY] = required;
		} else newPropertySchema.required = required;
		if (jsonSchema.$defs) newPropertySchema[require_symbols.DEFS_KEY] = jsonSchema.$defs;
		if (jsonSchema.id) newPropertySchema[require_symbols.PARENT_ID_KEY] = jsonSchema.id;
		if (typeof jsonSchema.additionalProperties === "boolean") newPropertySchema[require_symbols.PARENT_ADDITIONAL_PROPERTIES_KEY] = jsonSchema.additionalProperties;
		const reservedKeys = new Set([
			"type",
			"properties",
			"required",
			"additionalProperties",
			"$defs",
			"id"
		]);
		const parentMetadata = {};
		for (const [key, value] of Object.entries(jsonSchema)) if (!reservedKeys.has(key) && value !== void 0) parentMetadata[key] = value;
		if (Object.keys(parentMetadata).length > 0) newPropertySchema[require_symbols.PARENT_METADATA_KEY] = parentMetadata;
		properties[propertyKey] = newPropertySchema;
	}
	return properties;
}
function generateJsonSchema(schema, io) {
	const generatedJsonSchema = "_zod" in schema ? (0, zod_v4_core.toJSONSchema)(schema, {
		io,
		override: ({ jsonSchema }) => {
			if (io === "output" && "id" in jsonSchema) jsonSchema.id = `${jsonSchema.id}_Output`;
		}
	}) : require_symbols.zodV3ToOpenAPI(schema);
	const $defs = "$defs" in generatedJsonSchema && generatedJsonSchema.$defs ? generatedJsonSchema.$defs : void 0;
	const newSchema = fixRefsToPointById(generatedJsonSchema, $defs);
	const newDefs = {};
	Object.entries($defs || {}).forEach(([defKey, defValue]) => {
		if (defValue.id) {
			const newKey = defValue.id || defKey;
			if (newDefs[newKey]) throw new Error(`[nestjs-zod] Duplicate id in $defs: ${newKey}`);
			newDefs[newKey] = fixRefsToPointById(defValue, $defs);
		} else newDefs[defKey] = fixRefsToPointById(defValue, $defs);
	});
	if ($defs) newSchema.$defs = newDefs;
	return newSchema;
}
/**
* Changes all $refs in a schema to point based off the schema's ID, not the
* keys of the $defs object.  For example, given this schema:
* ```json
* {
*  type: "object",
*  properties: {
*    author: {
*      $ref: "#/$defs/Author"
*    }
*  }
* }
* ```
* and this $defs object:
* ```json
* {
*   Author: {
*    id: "Author_Output",
*    type: "object",
*    properties: {
*      // ...
*    }
*  }
* }
* ```
* It will return this schema:
* ```json
* {
*  type: "object",
*  properties: {
*    author: {
*      $ref: "#/$defs/Author_Output"
*    }
*  }
* }
* ```
*/
function fixRefsToPointById(rootSchema, $defs) {
	return require_symbols.walkJsonSchema(rootSchema, (schema) => {
		if (schema.$ref && schema.$ref.startsWith("#/$defs/")) {
			const defKey = schema.$ref.replace("#/$defs/", "");
			const defId = $defs?.[defKey].id;
			if (defId) schema.$ref = `#/$defs/${defId}`;
		}
		return schema;
	}, { clone: true });
}
function getSchemaMetadata(jsonSchema) {
	let hasRefs = false;
	let hasNull = false;
	let hasConst = false;
	require_symbols.walkJsonSchema(jsonSchema, (schema) => {
		if (schema.type === "null") hasNull = true;
		if (schema.$ref) hasRefs = true;
		if (schema.const) hasConst = true;
		return schema;
	});
	return {
		hasRefs,
		hasNull,
		hasConst
	};
}
function isZodDto(metatype) {
	return Boolean(metatype && (typeof metatype === "object" || typeof metatype === "function") && "isZodDto" in metatype && metatype.isZodDto);
}
function isObjectTypeWithProperties(jsonSchema) {
	return jsonSchema.type === "object" && !!jsonSchema.properties && Object.keys(jsonSchema.properties).length > 0;
}

//#endregion
exports.createZodDto = createZodDto;
exports.isZodDto = isZodDto;
