// nestjs-zod/dist/symbols-D0lN8Cd1.mjs
import deepmerge from "deepmerge";
import { z } from "zod/v3";

//#region src/assert.ts
function assert(condition, message = "Assertion failed") {
	if (!condition) throw new Error(`[nestjs-zod] ${message}`);
}

//#endregion
//#region src/const.ts
const PREFIX = "x-nestjs_zod";
const EMPTY_TYPE_KEY = `${PREFIX}-empty-type`;
const DEFS_KEY = `${PREFIX}-$defs`;
const PARENT_ID_KEY = `${PREFIX}-parent-id`;
const PARENT_ADDITIONAL_PROPERTIES_KEY = `${PREFIX}-parent-additional-properties`;
const PARENT_HAS_REFS_KEY = `${PREFIX}-parent-has-refs`;
const UNWRAP_ROOT_KEY = `${PREFIX}-unwrap-root`;
const HAS_NULL_KEY = `${PREFIX}-has-null`;
const HAS_CONST_KEY = `${PREFIX}-has-const`;
const SELF_REQUIRED_KEY = `${PREFIX}-self-required`;
const PARENT_METADATA_KEY = `${PREFIX}-parent-metadata`;

//#endregion
//#region src/utils.ts
function fixAllRefs({ schema, defRenames, rootSchemaName }) {
	return walkJsonSchema(schema, (s) => {
		if (s.$ref) {
			if (s.$ref.startsWith("#/$defs/")) {
				const oldDefName = s.$ref.replace("#/$defs/", "");
				const newDefName = defRenames?.[oldDefName];
				if (newDefName) s.$ref = `#/$defs/${newDefName}`;
			}
			s.$ref = s.$ref.replace("#/$defs/", "#/components/schemas/");
			if (s.$ref === "#") {
				if (!rootSchemaName) throw new Error("[fixAllRefs] rootSchemaName is required when fixing a ref to #");
				s.$ref = `#/components/schemas/${rootSchemaName}`;
			}
		}
		return s;
	}, { clone: true });
}
/**
* By default, zod generates openapi schemas that are compatible with OpenAPI
* 3.1.  But OpenAPI 3.0 supports a weird flavour of JSONSchema they call the
* "subset superset"
*
* This function converts the schema to the OpenAPI 3.0 subset superset format.
*
* See more information here:
* https://www.apimatic.io/blog/2021/09/migrating-to-and-from-openapi-3-1
*/
function convertToOpenApi3Point0(schema) {
	return walkJsonSchema(schema, (s) => {
		if ("id" in s) delete s.id;
		if (s.anyOf) {
			const nullSchema = s.anyOf.findIndex((subSchema) => subSchema.type === "null");
			if (nullSchema === -1) return s;
			s.anyOf.splice(nullSchema, 1);
			const { anyOf, ...rest } = s;
			if (anyOf.length === 1) {
				const sole = anyOf[0];
				if (sole.$ref && Object.keys(sole).length === 1) return {
					allOf: [sole],
					nullable: true,
					...rest
				};
				return {
					...sole,
					...rest,
					nullable: true
				};
			}
			return {
				...rest,
				anyOf,
				nullable: true
			};
		}
		if (typeof s.const !== "undefined") {
			s.enum = [s.const];
			delete s.const;
		}
		return s;
	}, { clone: true });
}
function walkJsonSchema(schema, callback, options) {
	schema = callback(options?.clone ? deepmerge(schema, {}) : schema);
	if (schema.type === "object" && schema.properties) for (const key in schema.properties) schema.properties[key] = walkJsonSchema(schema.properties[key], callback);
	if (schema.type === "array" && Array.isArray(schema.items)) schema.items = schema.items.map((item) => walkJsonSchema(item, callback));
	if (schema.type === "array" && schema.items) schema.items = walkJsonSchema(schema.items, callback);
	if (schema.oneOf) schema.oneOf = schema.oneOf.map((subSchema) => walkJsonSchema(subSchema, callback));
	if (schema.anyOf) schema.anyOf = schema.anyOf.map((subSchema) => walkJsonSchema(subSchema, callback));
	if (schema.allOf) schema.allOf = schema.allOf.map((subSchema) => walkJsonSchema(subSchema, callback));
	if (typeof schema.additionalProperties === "object") schema.additionalProperties = walkJsonSchema(schema.additionalProperties, callback);
	if (typeof schema.propertyNames === "object" && schema.propertyNames !== null) schema.propertyNames = walkJsonSchema(schema.propertyNames, callback);
	return schema;
}

//#endregion
//#region src/zodV3ToOpenApi.ts
function is(input, factory) {
	return factory === z[input._def.typeName];
}
/**
* @deprecated `zodToOpenAPI` will be removed in a future version, since zod
* v4 adds built-in support for generating OpenAPI schemas from zod schemas.
*/
function zodV3ToOpenAPI(zodType, visited = /* @__PURE__ */ new Set()) {
	const object = {};
	if (zodType.description) object.description = zodType.description;
	if (is(zodType, z.ZodString)) {
		const { checks } = zodType._def;
		object.type = "string";
		for (const check of checks) if (check.kind === "min") object.minLength = check.value;
		else if (check.kind === "max") object.maxLength = check.value;
		else if (check.kind === "email") object.format = "email";
		else if (check.kind === "url") object.format = "uri";
		else if (check.kind === "uuid") object.format = "uuid";
		else if (check.kind === "cuid") object.format = "cuid";
		else if (check.kind === "regex") object.pattern = check.regex.source;
		else if (check.kind === "datetime") object.format = "date-time";
	}
	if (is(zodType, z.ZodBoolean)) object.type = "boolean";
	if (is(zodType, z.ZodNumber)) {
		const { checks } = zodType._def;
		object.type = "number";
		for (const check of checks) if (check.kind === "int") object.type = "integer";
		else if (check.kind === "min") {
			object.minimum = check.value;
			object.exclusiveMinimum = !check.inclusive;
		} else if (check.kind === "max") {
			object.maximum = check.value;
			object.exclusiveMaximum = !check.inclusive;
		} else if (check.kind === "multipleOf") object.multipleOf = check.value;
	}
	if (is(zodType, z.ZodBigInt)) {
		object.type = "integer";
		object.format = "int64";
	}
	if (is(zodType, z.ZodArray)) {
		const { minLength, maxLength, type } = zodType._def;
		object.type = "array";
		if (minLength) object.minItems = minLength.value;
		if (maxLength) object.maxItems = maxLength.value;
		object.items = zodV3ToOpenAPI(type, visited);
	}
	if (is(zodType, z.ZodTuple)) {
		const { items } = zodType._def;
		object.type = "array";
		object.items = { oneOf: items.map((item) => zodV3ToOpenAPI(item, visited)) };
	}
	if (is(zodType, z.ZodSet)) {
		const { valueType, minSize, maxSize } = zodType._def;
		object.type = "array";
		if (minSize) object.minItems = minSize.value;
		if (maxSize) object.maxItems = maxSize.value;
		object.items = zodV3ToOpenAPI(valueType, visited);
		object.uniqueItems = true;
	}
	if (is(zodType, z.ZodUnion)) {
		const { options } = zodType._def;
		object.oneOf = options.map((option) => zodV3ToOpenAPI(option, visited));
	}
	if (is(zodType, z.ZodDiscriminatedUnion)) {
		const { options } = zodType._def;
		object.oneOf = [];
		for (const schema of options.values()) object.oneOf.push(zodV3ToOpenAPI(schema, visited));
	}
	if (is(zodType, z.ZodLiteral)) {
		const { value } = zodType._def;
		if (typeof value === "string") {
			object.type = "string";
			object.enum = [value];
		}
		if (typeof value === "number") {
			object.type = "number";
			object.minimum = value;
			object.maximum = value;
		}
		if (typeof value === "boolean") object.type = "boolean";
	}
	if (is(zodType, z.ZodEnum)) {
		const { values } = zodType._def;
		object.type = "string";
		object.enum = values;
	}
	if (is(zodType, z.ZodNativeEnum)) {
		const { values } = zodType._def;
		object.type = "string";
		object.enum = Object.values(values);
		object["x-enumNames"] = Object.keys(values);
	}
	if (is(zodType, z.ZodTransformer)) {
		const { schema } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(schema, visited));
	}
	if (is(zodType, z.ZodNullable)) {
		const { innerType } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(innerType, visited));
		object.nullable = true;
	}
	if (is(zodType, z.ZodOptional)) {
		const { innerType } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(innerType, visited));
	}
	if (is(zodType, z.ZodDefault)) {
		const { defaultValue, innerType } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(innerType, visited));
		object.default = defaultValue();
	}
	if (is(zodType, z.ZodObject)) {
		const { shape } = zodType._def;
		object.type = "object";
		object.properties = {};
		object.required = [];
		for (const [key, schema] of Object.entries(shape())) {
			object.properties[key] = zodV3ToOpenAPI(schema, visited);
			if (![z.ZodOptional.name, z.ZodDefault.name].includes(schema.constructor.name)) object.required.push(key);
		}
		if (object.required.length === 0) delete object.required;
	}
	if (is(zodType, z.ZodRecord)) {
		const { valueType } = zodType._def;
		object.type = "object";
		object.additionalProperties = zodV3ToOpenAPI(valueType, visited);
	}
	if (is(zodType, z.ZodIntersection)) {
		const { left, right } = zodType._def;
		const merged = deepmerge(zodV3ToOpenAPI(left, visited), zodV3ToOpenAPI(right, visited), { arrayMerge: (target, source) => {
			const mergedSet = new Set([...target, ...source]);
			return Array.from(mergedSet);
		} });
		Object.assign(object, merged);
	}
	if (is(zodType, z.ZodEffects)) {
		const { schema } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(schema, visited));
	}
	if (is(zodType, z.ZodLazy)) {
		const { getter } = zodType._def;
		if (visited.has(getter)) return object;
		visited.add(getter);
		Object.assign(object, zodV3ToOpenAPI(getter(), visited));
	}
	return object;
}

//#endregion
//#region src/symbols.ts
const ioSymbol = Symbol("io");

//#endregion
export { walkJsonSchema as a, HAS_CONST_KEY as c, PARENT_HAS_REFS_KEY as d, PARENT_ID_KEY as f, assert as g, UNWRAP_ROOT_KEY as h, fixAllRefs as i, HAS_NULL_KEY as l, SELF_REQUIRED_KEY as m, zodV3ToOpenAPI as n, DEFS_KEY as o, PARENT_METADATA_KEY as p, convertToOpenApi3Point0 as r, EMPTY_TYPE_KEY as s, ioSymbol as t, PARENT_ADDITIONAL_PROPERTIES_KEY as u };
